### ISSUE TITLE
On demand data transfers in the mesh

### ISSUE THREAD

**Issue by labkode** [sciencemesh/sciencemesh#51]
As part of task 4.4 and the integration of the on-demand data transfers in the IOP we involved Eddie from CERN FTS for this use-case.

![fts](https://user-images.githubusercontent.com/5175912/79726985-f5252880-82eb-11ea-8353-98ea626d877b.png)

The IOP will be deployed at all the sites and the place where the development needs to happen to add the on-demand data transfers.

We consider in this use case to explore FTS as it is a known technology at the partners sites.
In this example, we have two sites: CERN and AARNET and both have the IOP.

A data transfer service, in this case FTS will be deployed as part of the sciencemesh project.

How to authenticate a transfer from CERN to AARNET?

* A) User Hugo logs-in at CERN site and gets an access token, the CERN IOP contact FTS to organise the transfer, FTS streams the data from CERN to AARNET and use the same access token at AARNET site. This approach is very broad as local AAI technologies can range from simple SAML to OIDC, so it needs to be investigated at the AAI level for token delegation/token re-use fro third-party services (like FTS). @david-antos 

* B) User Hugo logs-in CERN and initiates a data transfer. FTS is a trusted service for both IOPs to impersonate users at both sides. The CERN IOP contacts FTS, FTS streams data from source to destination.

* C) User Hugo logs-in at CERN site an initiates a data transfer.  The IOP forges a constrained token for the resources the user wants to transfer and gives them to FTS. FTS tell AARNET to initiate a WebDAV third-party copy and AARNET pulls the files using the forged token at CERN site. This is most promising solution for two reasons: 1) most efficient as data is not streamed 2) access is independent of local AAI. Requires input from @EdwardKaravakis on how FTS notifies the target system to start pulling. I think maybe using `webdav COPY` method. @EdwardKaravakis can you link the Webdav 3rd-party copy spec here?

To initiate the transfer we'll piggy back on OCM, extending the `protocol` to let the target system of a data transfer being requested so the user can confirm the transfer and monitor the transfer in a background job.

The request for the data transfer UI is vendor-dependant.

**Comment by EdwardKaravakis** [comment]
Absolutely, please find the protocol-level specs here: https://twiki.cern.ch/twiki/bin/view/LCG/HttpTpcTechnical (not sure if it is available for everyone though)

**Comment by labkode** [comment]
PLAIN REVA TRANSFERS
==============
* Extend OCM WebDAV protocol to pull data from source instead of online access
* When invitation is accepted, implement in Reva background process to fetch the data and store in user home directory

FILESENDER
========
Individual can use FileSender independently if they have a EFSS or not.

* FileSender implements OCM share creation: a user of FileSender will have a button to send data to an EFSS like a normal EFSS will do.
